### DISCUSSION TITLE
Avoid using MA27 in BlockSQP

### DISCUSSION THREAD

**Discussion by itbellix** [casadi/casadi#3546] [Q&A]
Hi all,
I have always used Ipopt as a solver, through CasADi, but I am trying to experiment with other (hopefully faster) solvers. Note that Ipopt works fine (converges to a solution) on the problem that I am considering.

I have seen that CasADi distributes a (slightly modified?) version  of the [blockSQP](https://github.com/djanka2/blockSQP) solver, and I would be interested to try it. However, when I do so, I get the following error:

```
+---------------------------------------------------------------+
| Starting blockSQP with the following algorithmic settings:    |
+---------------------------------------------------------------+
| qpOASES flavor            | sparse, Schur complement approach |
| Globalization             | filter line search                |
| 1st Hessian approximation | block L-SR1, OL                   |
| 2nd Hessian approximation | block L-BFGS, selective sizing    |
+---------------------------------------------------------------+

   it    qpIt                obj      feas       opt    |lgrd|     |stp|    |lstp|    alpha   nSOCS sk, da, sca       QPr,mu
    0            0  2.434771e+05  2.77e+01  1.23e+05  
python3: symbol lookup error: /home/itbellix/.local/lib/python3.8/site-packages/casadi/libcasadi_linsol_ma27.so: undefined symbol: ma27id_
```

I have seen another post (https://github.com/casadi/casadi/issues/2311) where the same issue is discussed. There the library `libhsl.so` (from HSL) appeared to be the problem. 

However, I would like my code to be as user-friendly as possible, so I would like to know if there is a way to use another linear solver for blockSQP in CasADi (like MUMPS that is used by Ipopt by default). In this way, simply installing CasADi with `pip` would already allow the user to run the optimization.

Is there a way to do this? Or it does not make sense for any reason?

Thank you!

**Comment by sandeep026** [comment]
It your problem is sparse, then using MA27 will be of help. By default, I think the sparse variant of qpoases is used for which ma27 is required. IIRC, there is an option called 'schur'  which can be turned into false use the dense mode of qpoases. Hope that helps.

I had luckily managed to install ma27 on previous machine.

**Reply by sandeep026** [reply]
A presentation on sparse qpoases. 
https://www.syscop.de/files/2014ss/imtek-tempo/EQP_2014_Kirches.pdf

**Reply by itbellix** [reply]
Hi @sandeep026, thank you for your answer. I was not aware that `qpOASES`and `blockSQP` are leveraging by default MA27 under the hood. Indeed, this is quite clear in [Blocksqp::init](https://github.com/casadi/casadi/blob/ddd68adcdc2fa9a1ebfc30b5c7b3610eb3040e60/casadi/interfaces/blocksqp/blocksqp.cpp#L238).

I was hoping that MUMPS (which is also a sparse solver) could also be used by `blockSQP` in a similar manner. The answer to this appears to be no, as I did not find any option to set the linear solver to be different that the default one...

I have tried setting `schur=False`. For some reason `blockSQP` does not converge in this way (on the exact same problem IPOPT does), and then still complains that it cannot find `libcasadi_linsol_ma27.so`. I think that what is happening is that `blockSQP` tries to switch to the default options, where it indeed needs MA27.

**Reply by sandeep026** [reply]
I ran the example script and it did not report any errors about ma27.  Unless a good initialization is provided, free solvers other than IPOPT generally have a hard time converging to a solution.

```
import casadi as cs
import matplotlib.pyplot as plt
opti = cs.Opti()

x = opti.variable()
y = opti.variable()

opti.minimize((1-x)**2+(y-x**2)**2)

opti.solver('blocksqp',{'schur':False})
sol = opti.solve()

plt.plot(sol.value(x),sol.value(y),'o')
```

```
***WARNING: SR1 update only works with qpOASES Schur complement version. Using BFGS updates instead.

+---------------------------------------------------------------+
| Starting blockSQP with the following algorithmic settings:    |
+---------------------------------------------------------------+
| qpOASES flavor            | sparse, reduced Hessian factorization|
| Globalization             | filter line search                |
| 1st Hessian approximation | block L-BFGS, selective sizing    |
| 2nd Hessian approximation | -                                 |
+---------------------------------------------------------------+

   it    qpIt                obj      feas       opt    |lgrd|     |stp|    |lstp|    alpha   nSOCS sk, da, sca       QPr,mu
    0            0  1.000000e+00  0.00e+00  2.00e+00  
    1      4+    0  3.125000e-01  0.00e+00  5.00e-01  5.00e-01  5.00e-01  0.00e+00  2.5e-01      0  0,   0, 0.0e+00  0, 0.0e+00  
    2      1+    0  2.758726e-02  0.00e+00  2.47e-01  2.47e-01  6.67e-01  0.00e+00  1.0e+00      0  0,   0, 1.0e+00  0, 0.0e+00  
    3      1+    0  2.290177e-02  0.00e+00  6.17e-01  6.17e-01  2.71e-01  0.00e+00  1.0e+00      0  0,   0, 9.7e-01  0, 0.0e+00  
    4      1+    0  1.682397e-04  0.00e+00  1.53e-02  1.53e-02  8.42e-02  0.00e+00  1.0e+00      0  0,   0, 1.0e+00  0, 0.0e+00  
    5      1+    0  9.287317e-06  0.00e+00  5.25e-03  5.25e-03  2.22e-02  0.00e+00  1.0e+00      0  0,   0, 1.0e+00  0, 0.0e+00  
    6      1+    0  2.394713e-08  0.00e+00  6.43e-04  6.43e-04  6.15e-03  0.00e+00  1.0e+00      0  0,   0, 9.8e-01  0, 0.0e+00  
    7      1+    0  5.074628e-10  0.00e+00  9.13e-05  9.13e-05  1.49e-04  0.00e+00  1.0e+00      0  0,   0, 1.0e+00  0, 0.0e+00  
    8      1+    0  1.942930e-17  0.00e+00  1.96e-08  1.96e-08  2.13e-05  0.00e+00  1.0e+00      0  0,   0, 1.0e+00  0, 0.0e+00  

***CONVERGENCE ACHIEVED!***
      solver  :   t_proc      (avg)   t_wall      (avg)    n_eval
      nlp_fg  |        0 (       0)  19.00us (  1.46us)        13
   nlp_gf_jg  |        0 (       0)  12.00us (  1.33us)         9
       total  |   5.00ms (  5.00ms)   5.19ms (  5.19ms)         1

```

Additionally I tried to  pass the options {'schur':False,'sparse':False} for qpsol_options, but casadi reports that it is not supported.
For sqpmethod it did not report this.

```
CasADi - 2024-01-23 15:13:36 WARNING("Option 'qpsol_options' currently not supported, ignored") [.../casadi/interfaces/blocksqp/blocksqp.cpp:300]
```
